File: src/main/java/org/radarcns/stream/phone/PhoneUsageCollector.java
package org.radarcns.stream.phone;

import org.radarcns.passive.phone.PhoneUsageEvent;
import org.radarcns.passive.phone.UsageEventType;

import java.math.BigDecimal;
import java.math.MathContext;

/**
 * Created by piotrzakrzewski on 27/07/2017.
 */
public class PhoneUsageCollector {
    private BigDecimal totalForegroundTime; // total time in seconds
    private double lastForegroundEvent; // date in Unix time in seconds
    private int timesTurnedOn;
    private String categoryName; // optional
    private Double categoryNameFetchTime; // optional

    public PhoneUsageCollector() {
        totalForegroundTime = BigDecimal.ZERO;
    }

    public PhoneUsageCollector update(PhoneUsageEvent event) {
        if (event.getCategoryName() != null) {
            this.categoryName = event.getCategoryName();
            this.categoryNameFetchTime = event.getCategoryNameFetchTime();
        }

        if (event.getEventType() == UsageEventType.FOREGROUND) {
            // Foreground event received
            timesTurnedOn++;
            lastForegroundEvent = event.getTime();
        } else if (event.getEventType() == UsageEventType.BACKGROUND
                && lastForegroundEvent != 0.0) {
            // Background event received for an app which was previously on.
            totalForegroundTime = totalForegroundTime.add(
                    BigDecimal.valueOf(event.getTime())
                            .subtract(BigDecimal.valueOf(lastForegroundEvent),
                                    MathContext.DECIMAL128));
            lastForegroundEvent = 0.0;
        }
        // else if eventType is background and it was already in the background, ignore.
        // We must have missed an event.
        // else if eventType is something else: ignore.

        return this;
    }

    public double getTotalForegroundTime() {
        return totalForegroundTime.doubleValue();
    }

    public void setTotalForegroundTime(double totalForegroundTime) {
        this.totalForegroundTime = BigDecimal.valueOf(totalForegroundTime);
    }

    public double getLastForegroundEvent() {
        return lastForegroundEvent;
    }

    public void setLastForegroundEvent(double lastForegroundEvent) {
        this.lastForegroundEvent = lastForegroundEvent;
    }

    public int getTimesTurnedOn() {
        return timesTurnedOn;
    }

    public void setTimesTurnedOn(int timesTurnedOn) {
        this.timesTurnedOn = timesTurnedOn;
    }

    public String getCategoryName() {
        return categoryName;
    }

    public void setCategoryName(String categoryName) {
        this.categoryName = categoryName;
    }

    public Double getCategoryNameFetchTime() {
        return categoryNameFetchTime;
    }

    public void setCategoryNameFetchTime(Double categoryNameFetchTime) {
        this.categoryNameFetchTime = categoryNameFetchTime;
    }
}

File: src/main/kotlin/org/radarbase/stream/phone/PhoneUsageCollector.kt
package org.radarbase.stream.phone

import org.radarcns.passive.phone.PhoneUsageEvent
import org.radarcns.passive.phone.UsageEventType
import java.math.BigDecimal
import java.math.MathContext

/**
 * Created by piotrzakrzewski on 27/07/2017.
 * Converted to Kotlin.
 */
class PhoneUsageCollector {
    var totalForegroundTime: BigDecimal = BigDecimal.ZERO
        private set

    var lastForegroundEvent: Double = 0.0
    var timesTurnedOn: Int = 0
    var categoryName: String? = null
    var categoryNameFetchTime: Double? = null

    /**
     * Updates the collector state based on a new [PhoneUsageEvent].
     */
    fun update(event: PhoneUsageEvent): PhoneUsageCollector {
        // Update category info if present
        event.categoryName?.let {
            this.categoryName = it
            this.categoryNameFetchTime = event.categoryNameFetchTime
        }

        when (event.eventType) {
            UsageEventType.FOREGROUND -> {
                timesTurnedOn++
                lastForegroundEvent = event.time
            }
            UsageEventType.BACKGROUND -> {
                if (lastForegroundEvent != 0.0) {
                    val duration = BigDecimal.valueOf(event.time)
                        .subtract(BigDecimal.valueOf(lastForegroundEvent), MathContext.DECIMAL128)
                    
                    totalForegroundTime = totalForegroundTime.add(duration)
                    lastForegroundEvent = 0.0
                }
            }
            else -> {
                // Ignore other event types or repeated backgrounds
            }
        }

        return this
    }

    // Helper property to match the original Java double getter/setter logic
    var totalForegroundTimeDouble: Double
        get() = totalForegroundTime.toDouble()
        set(value) {
            totalForegroundTime = BigDecimal.valueOf(value)
        }
}